index.js

```diff
@@ -0,0 +1,130 @@
+import { IPLDURL } from 'js-ipld-url'
+import { create as createTyped } from '@ipld/schema/typed.js'
+import { toDSL } from '@ipld/schema/to-dsl.js'
+import { CID } from 'multiformats/cid'
+import { base32 } from 'multiformats/bases/base32'
+import { base36 } from 'multiformats/bases/base36'
+
+export const DEFAULT_CID_BASES = base32.decoder.or(base36.decoder)
+
+export default class IPLDURLSystem {
+  constructor ({
+    getNode,
+    adls = new Map(),
+    cidBases = DEFAULT_CID_BASES
+  }) {
+    if(!getNode) throw new TypeError('Must provide a getNode function')
+    this.getNode = getNode
+    this.adls = adls
+    this.cidBases = cidBases
+  }
+
+  async resolve (url) {
+    const { hostname: root, segments, searchParams } = new IPLDURL(url)
+
+    const cid = CID.parse(root, this.cidBases).toV1()
+    let data = await this.getNode(cid)
+
+    const initialParameters = {}
+    let shouldProcessRoot = false
+    if (searchParams.has('schema')) {
+      shouldProcessRoot = true
+      initialParameters.schema = searchParams.get('schema')
+      initialParameters.type = searchParams.get('type')
+    }
+    if (searchParams.has('adl')) {
+      // TODO Should other parameters be passed to the ADL function?
+      shouldProcessRoot = true
+      initialParameters.adl = searchParams.get('adl')
+    }
+    if (shouldProcessRoot) {
+      data = await this.#applyParameters(data, initialParameters)
+    }
+
+    // TODO: Account for dag-pb?
```

Comment:
for named links? I reckon not, there's a lot of regret about that now because it feels like special-casing that you have to escape out of with an `/ipld` prefix, but since this is all IPLD then how about not special-casing at all _or_ making it an opt-in with an ADL!

Reply:
Cool, lets pretend dag-pb doesn't exist for this case. If folks wanna path over it with URLs, they can either use an ADL or use `/Link/0/` to traverse the parsed data.